src/modules/ekf2/EKF/zero_gyro_update.cpp

```diff
@@ -0,0 +1,89 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2023 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file zero_gyro_update.cpp
+ * Control function for ekf zero gyro update
+ */
+
+#include "ekf.h"
+
+void Ekf::controlZeroGyroUpdate(const imuSample &imu_delayed)
+{
+	// Downsample gyro data to run the fusion at a lower rate
+	_zgup_delta_ang += imu_delayed.delta_ang;
+	_zgup_delta_ang_dt += imu_delayed.delta_ang_dt;
+
+	static constexpr float zgup_dt = 0.2f;
+	const bool zero_gyro_update_data_ready = _zgup_delta_ang_dt >= zgup_dt;
+
+	if (zero_gyro_update_data_ready) {
+		const bool continuing_conditions_passing = _control_status.flags.vehicle_at_rest
+				&& _control_status_prev.flags.vehicle_at_rest;
+
+		if (continuing_conditions_passing) {
+			Vector3f delta_ang_scaled = _zgup_delta_ang / _zgup_delta_ang_dt * _dt_ekf_avg;
```

Comment:
I'm wondering if we should start using the specific dt everywhere appropriate? Maybe something to discuss later after accel/gyro bias states.

```suggestion
			Vector3f delta_ang_scaled = _zgup_delta_ang / _zgup_delta_ang_dt * imu_delayed.delta_ang_dt;
```

Reply:
The filter estimates the delta angle based on `_dt_ekf_avg`, this is why we need to scale it with that value here. But yes, I generally agree that using angular rate values will reduce complexity.